url11fb0973cf_raichuman9_Swimmer526_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/fr/inria/optimization/cmaes/CMAEvolutionStrategy.java

```diff
@@ -0,0 +1,2704 @@
+     * @return y
+     */
+    double[] phenoGenoTransformation(double[] x, double[] y) {
+    	if (y == null || y == x || y.length != x.length) {
+    		y = x.clone();
+    		return y; // for now return an identical copy
+    	}
+    	for(int i = 0; i < N; ++i)
+    		y[i] = x[i];
+    	return y;		
+    }
+    
+    /**
+     * Samples the recent search distribution lambda times
+     * @return double[][] population, lambda times dimension array of sampled solutions, 
+     *   where <code>lambda == parameters.getPopulationSize()</code> 
+     * @see #resampleSingle(int)
+     * @see #updateDistribution(double[])
+     * @see CMAParameters#getPopulationSize()
+     */
+    public double[][] samplePopulation() {
+        int i, j, iNk;
+        double sum;
+
+        if (state < 0)
+            init();
+        else if (state < 3 && state > 2)
+                error("mixing of calls to updateSingle() and samplePopulation() is not possible");
+        else    
+            eigendecomposition(0); // latest possibility to generate B and diagD
+        
+        if (state != 1)
+            ++countiter; 
+        state = 1; // can be repeatedly called without problem
+        idxRecentOffspring = sp.getLambda() - 1; // not really necessary at the moment
+
+        
+        // ensure maximal and minimal standard deviations
+        if (options.lowerStandardDeviations != null && options.lowerStandardDeviations.length > 0)
+            for (i = 0; i < N; ++i) {
+                double d = options.lowerStandardDeviations[Math.min(i,options.lowerStandardDeviations.length-1)]; 
+                if(d > sigma * minsqrtdiagC) 
+                    sigma = d / minsqrtdiagC;
+            }
+        if (options.upperStandardDeviations != null && options.upperStandardDeviations.length > 0)
+            for (i = 0; i < N; ++i) {
+                double d = options.upperStandardDeviations[Math.min(i,options.upperStandardDeviations.length-1)]; 
+                if (d < sigma * maxsqrtdiagC) 
+                    sigma = d / maxsqrtdiagC;
+            }
+        
+        testAndCorrectNumerics();
+        
+        /* sample the distribution */
+        for (iNk = 0; iNk < sp.getLambda(); ++iNk) { /*
+            * generate scaled
+            * random vector (D * z)
+            */
+
+            // code duplication from resampleSingle because of possible future resampling before GenoPheno
+            /* generate scaled random vector (D * z) */
+        	if (flgdiag)  
+        	    for (i = 0; i < N; ++i)
+        			arx[iNk][i] = xmean[i] + sigma * diagD[i] * rand.nextGaussian();
+        	else {
+                for (i = 0; i < N; ++i) 
+        		    artmp[i] = diagD[i] * rand.nextGaussian();
+        	
+        	    /* add mutation (sigma * B * (D*z)) */
+        	    for (i = 0; i < N; ++i) {
+        			for (j = 0, sum = 0; j < N; ++j)
+        				sum += B[i][j] * artmp[j];
+        			arx[iNk][i] = xmean[i] + sigma * sum;
+        		}
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1513:1523, 1557:1567.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.